symforce/slam/imu_preintegration/manifold_symbolic.py

```diff
@@ -0,0 +1,346 @@
+
+    return upper_triangular_mul(sqrt_info, res)
+
+
+def right_jacobian(tangent: sf.V3, epsilon: T.Scalar) -> sf.M33:
+    """
+    The right jacobian J(v) is d/du Log(Exp(v)^{-1} Exp(v + u)), i.e., a matrix such that
+    Exp(v + dv) ~= Exp(v) Exp(J(v) dv), where v and u are tangent vectors of SO(3).
+
+    Returns J(tangent).
+    """
+    norm = tangent.norm(epsilon)
+    tangent_hat = sf.Matrix.skew_symmetric(tangent)
+    out = sf.M33.eye()
+    out += ((1 - sf.cos(norm)) / (norm ** 2)) * tangent_hat
+    out += ((norm - sf.sin(norm)) / (norm ** 3)) * (tangent_hat * tangent_hat)
+    return out
+
+
+def handwritten_state_D_old_state_gyro_accel(
+    DR: sf.Rot3, corrected_gyro: sf.V3, corrected_accel: sf.V3, dt: T.Scalar, epsilon: T.Scalar
+) -> T.Tuple[sf.Matrix, sf.Matrix, sf.Matrix]:
+    """
+    Calculates the derivatives of the new state (meaning the DR, Dv, and Dp) w.r.t. the previous state,
+    gyroscope measurement, and the accelerometer mesaurement.
+
+
+    Args:
+        DR (sf.Rot3): Preintegrated DR of the previous state
+        corrected_gyro (sf.V3): gyroscope measurment corrected for IMU bias
+        corrected_accel (sf.V3): accelerometer measurement corrected for IMU bias
+        dt (T.Scalar): Time difference between the previous time step and the new time step
+        epsilon (T.Scalar): epsilon for numerical stability
+
+    Returns:
+        T.Tuple[sf.M99, sf.M93, sf.M93]: new_state_D_old_state, new_state_D_gyro_measurement, new_state_D_accel_measurement
+    """
+    # NOTE(Brad): I put each of the 3 derivatives in one function to make it easier to keep
+    # subexpressions similar to improve CSE
+
+    # calculate new_state_D_old_state
+    DR_update = sf.Rot3.from_tangent(corrected_gyro * dt, epsilon)
+    new_DR_D_old_DR = DR_update.inverse().to_rotation_matrix()
+    new_DR_D_old_Dv = sf.M33.zero()
+    new_DR_D_old_Dp = sf.M33.zero()
+
+    new_Dv_D_old_DR = -(DR.to_rotation_matrix() * sf.Matrix.skew_symmetric(corrected_accel * dt))
+    new_Dv_D_old_Dv = sf.M33.eye()
+    new_Dv_D_old_Dp = sf.M33.zero()
+
+    new_Dp_D_old_DR = (
+        (-(DR.to_rotation_matrix() * sf.Matrix.skew_symmetric(corrected_accel * dt))) * dt / 2
+    )
+    new_Dp_D_old_Dv = dt * sf.M33.eye()
+    new_Dp_D_old_Dp = sf.M33.eye()
+
+    new_state_D_old_state = sf.Matrix.block_matrix(
+        [
+            [new_DR_D_old_DR, new_DR_D_old_Dv, new_DR_D_old_Dp],
+            [new_Dv_D_old_DR, new_Dv_D_old_Dv, new_Dv_D_old_Dp],
+            [new_Dp_D_old_DR, new_Dp_D_old_Dv, new_Dp_D_old_Dp],
+        ]
+    )
+
+    # calculate new_D_gyro
+    new_DR_D_gyro = right_jacobian(corrected_gyro * dt, epsilon) * dt
+    new_DvDp_D_gyro = sf.M63.zero()
+
+    new_state_D_gyro = sf.Matrix.block_matrix(
+        [
+            [new_DR_D_gyro],
+            [new_DvDp_D_gyro],
+        ]
+    )
+
+    # calculate new_D_accel
+    new_DR_D_accel = sf.M33.zero()
+    new_Dv_D_accel = DR.to_rotation_matrix() * dt
```

Comment:
Again, please double-check the sign here. This is saying that if the bias accel increases by `x`, then the preintegrated velocity should increase by `x * dt`. However, during integration, you actually subtract the bias, so I think there should be a negative sign somewhere.

Reply:
Yeah, I realized that after reading a bit more down the line.